### PR TITLE
Add cibuildwheel action from mpl

### PR DESCRIPTION
**Pull request by jni**
## Description

@emmanuelle here's a start! 😊 Don't forget to delete the branch after merging!

@hmaarrfk if you have any comments on this they would sure be appreciated!!! 🙏 

## Checklist

- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

**Comment by emmanuelle** [comment]
We probably want to run tests using the wheels after creating them.

**Comment by hmaarrfk** [comment]
I know it is tempting to merge the two repos, but there is a real concern in debugging the wheel builder, which could give you version 0.18.5 pretty quickly once it bugs out on releasing 0.18.0.

Is that something we are OK with?

I mean, I would be happy if we:

1. Forked `towncrier` because it hasn't been released in a while: https://github.com/twisted/towncrier/issues/269
2. Start using towncrier.
3. Start releasing more often.

But if we aren't going to release more often, maybe stick with the wheel builder repo?

**Comment by emmanuelle** [comment]
@hmaarrfk could you please elaborate a bit more? We started discussing this in #5045, but I don't understand why we can't have a config file for testing the package and another one for building the wheels, which can have different compilers options, versions of numpy etc. Please tell me what I'm missing :-).

**Comment by emmanuelle** [comment]
@jni the wheel building part is now working (I had to separate Mac for some specific compiler options), I've started adding CIBW env variables in order to test the wheels.

**Comment by jni** [comment]
@emmanuelle, I think @hmaarrfk's concern is that if you tag the repo, and it builds but fails, you then have to tag again to try again. But that happened to us on the other repo anyway, and I think that concern can be addressed quite effectively with release candidates.

Additionally, @hmaarrfk I know I was originally skeptical of your frequent releases proposal, but having experienced a year of doing it with napari, I'm a convert! So I absolutely want to use this infrastructure to start doing more frequent releases.

**Comment by jni** [comment]
Additionally, @hmaarrfk, I would say that I'm pretty ok with 0.18.5 as a concept. LOL

**Comment by jni** [comment]
This test seems a bit too stringent:

https://github.com/scikit-image/scikit-image/pull/5080/checks?check_run_id=1456068555#step:6:910

```pytb
    @test_parallel(warnings_matching=['Possible precision loss'])
    def check():
        expected = self.refs_3d[filter]
        result = getattr(rank, filter)(self.volume, self.selem_3d)
>       assert_array_equal(expected, result)
E       AssertionError: 
E       Arrays are not equal
E       
E       Mismatched elements: 13 / 1000 (1.3%)
E       Max absolute difference: 4.4408921e-16
E       Max relative difference: 1.7611122e-16
E        x: array([[[2.      , 2.321928, 2.321928, 2.321928, 2.321928, 2.321928,
E                2.321928, 2.321928, 2.321928, 2.      ],
E               [2.321928, 2.584963, 2.584963, 2.584963, 2.584963, 2.584963,...
E        y: array([[[2.      , 2.321928, 2.321928, 2.321928, 2.321928, 2.321928,
E                2.321928, 2.321928, 2.321928, 2.      ],
E               [2.321928, 2.584963, 2.584963, 2.584963, 2.584963, 2.584963,...
```

😂 I'll update to almost_equal...

**Comment by jni** [comment]
Ok we are in business people!!! 

Screen Shot 2020-11-26 at 12 39 55 pm

Screen Shot 2020-11-26 at 12 36 40 pm

Anyone willing to push the green button? =D

There are various questions, such as whether we want to build manylinux1 vs manylinux2010 wheels, but these should be better addressed in follow-up PRs, I think!

File: .github/workflows/cibuildwheel.yml
Comment (hmaarrfk) [inline .github/workflows/cibuildwheel.yml:42]: what is this? Where does it get used?

File: .github/workflows/cibuildwheel.yml
Comment (hmaarrfk) [inline .github/workflows/cibuildwheel.yml:92]: Why do we install numpy first? We have things pinned in pyproject.toml ;) don't we?

File: .github/workflows/cibuildwheel.yml
Comment (hmaarrfk) [inline .github/workflows/cibuildwheel.yml:98]: I thought we were dropping this?

File: .github/workflows/cibuildwheel.yml
Comment (jni) [reply .github/workflows/cibuildwheel.yml:42]: manylinux1 is the type of Linux wheel that gets produced (see my comment above). I guess [cibuildwheel](https://cibuildwheel.readthedocs.org) uses this env variable to determine how to build the wheels.

File: .github/workflows/cibuildwheel.yml
Comment (hmaarrfk) [inline .github/workflows/cibuildwheel.yml:33]: This is amazing!! Why are we pinned at 1.6.3 though?

File: .github/workflows/cibuildwheel.yml
Comment (jni) [reply .github/workflows/cibuildwheel.yml:98]: Yes but @emmanuelle wasn't quite committed to that, partly because of your earlier skepticism in #5053. I would be happy to drop this. I'd still like to get @emmanuelle's blessing though, so perhaps that can happen in a followup PR?

File: .github/workflows/cibuildwheel.yml
Comment (jni) [reply .github/workflows/cibuildwheel.yml:33]: That's because matplotlib pinned it 😂 

https://github.com/matplotlib/matplotlib/blob/2303c2b4afa7ecd1c3c576377c77dc53e0c86273/.github/workflows/cibuildwheel.yml

I'd like to take an "if it ain't broke" approach here for now. We can become adventurous after 0.18?

File: .github/workflows/cibuildwheel.yml
Comment (jni) [reply .github/workflows/cibuildwheel.yml:92]: In theory! I'd love to change this in a follow-up PR though, because the testing cycle is so long on changes here. Perhaps raise an issue with these tweaks to the build process to investigate once this is in?

File: .github/workflows/cibuildwheel.yml
Comment (jni) [reply .github/workflows/cibuildwheel.yml:98]: Actually, scrap that: this particular section is only triggered on 0.17 (see IF below). So I think it's appropriate to keep it.

Comment (jni) [reply .github/workflows/cibuildwheel.yml:98]: You can see in my screenshot that no 3.6 wheels were built.

File: .github/workflows/cibuildwheel.yml
Comment (hmaarrfk) [reply .github/workflows/cibuildwheel.yml:33]: ok

Comment (emmanuelle) [reply .github/workflows/cibuildwheel.yml:33]: `1.6.4` is the most recent version and `1.6.3` dates back to Oct 12, so quite recent.

**Comment by emmanuelle** [comment]
Regarding tags, I have to admit that for 0.17 I had to delete tags because of such problems. My approach is that we should not touch a tag once files have been pushed to pypi but that it's not so bad to change the tag before that. But of course I realize this is not completely orthodox, so this needs to be discussed!

**Comment by emmanuelle** [comment]
Also as it is written now, wheels will be built when new commits are pushed to master or release branches. Maybe we can tag after checking that the build went smoothly?

**Comment by emmanuelle** [comment]
@jni not all builds were tested so I added the environment variables for testing all wheels. Also I tried to remove this strange installation of a pinned numpy and of certify, hopefully we don't need it.

**Comment by jni** [comment]
Ok thanks @emmanuelle. One concern I have is that this is a lot of tests now, might we hit gh-actions timeout limits? I also would like to consider testing the wheels in a totally different action, or after the artifact is uploaded. It's frustrating at release time to be blocked by things that you know are unrelated, e.g. failure of pooch to download some data.

**Comment by emmanuelle** [comment]
Hum it's quite long indeed. With tests being configured in environment variables I'm not sure it's easy to have the tests run after uploading artifacts. However, probably we could have more jobs in parallel by having the python version as an element of the matrix (I'm not sure why py3.9 is separated from py 3.8 and 3.7 at the moment, it seems to be the same configuration).

Also when do we want to build the wheels: not on every commit I guess? On release tags? On pushes to master and release branches? (that's what scikit-learn does)

**Comment by jni** [comment]
> Also when do we want to build the wheels: not on every commit I guess? On release tags? On pushes to master and release branches? (that's what scikit-learn does)

I think this is already properly configured: on tags, pushes to master, pushes to release branches, and pushes to the "cibuildwheel" branch, which we can use whenever we want to make changes.

> Hum it's quite long indeed.

Personally I think it's overkill, but I don't mind leaving it for now, getting this merged, getting 0.18 out, and then tweaking for speed. I think we have done enough fiddling that I hope 0.18 will be quite smooth now!

**Comment by jni** [comment]
@scikit-image/core anyone up for some approvals/merges? ;)

**Comment by grlee77** [comment]
> Regarding tags, I have to admit that for 0.17 I had to delete tags because of such problems. My approach is that we should not touch a tag once files have been pushed to pypi but that it's not so bad to change the tag before that. But of course I realize this is not completely orthodox, so this needs to be discussed!

I have also overwritten tags in this manner for PyWavelets before (always before pushing anything to PyPI and typically within a short time frame of < 1 hour from tag creation).

File: .github/workflows/cibuildwheel.yml
Comment (grlee77) [reply .github/workflows/cibuildwheel.yml:42]: It seems we probably do want to add more recent manylinux builds fairly soon. The following recent blog post may be useful: https://labs.quansight.org/blog/2020/11/manylinux1-is-obsolete-manylinux2010-is-almost-eol-what-is-next/

**Review by grlee77 — APPROVED** [review]
I don't have GitHub Actions expertise or `cibuildwheel` experience, but the script is pretty understandable and the wheel artifacts look correct to me.

**Comment by grlee77** [comment]
The documentation failure looks like a timeout that can be ignored

**Comment by alexdesiqueira** [comment]
I'll try re-running it @grlee77.

**Comment by alexdesiqueira** [comment]
@grlee77 I'll try re-running it.

**Comment by emmanuelle** [comment]
The CircleCI failure is indeed quite common, unfortunately. It would be worth investigating it but it's not related at all to the PR.

**Comment by alexdesiqueira** [comment]
@jni :joy:

**Comment by jni** [comment]
Not sure if you were waiting for AppVeyor but I think that queue is like 700 builds long at this point so I wasn't about to. 😂

**Comment by alexdesiqueira** [comment]
I was still waiting for it. Good call, then :joy:

**Pull request merged**
